common/tools/dev-tool/src/framework/check.ts

```diff
@@ -0,0 +1,297 @@
+          cwd: project.path,
+          captureExitCode: true,
+          stdio: verbose ? "inherit" : undefined,
+          captureOutput: true,
+        });
+        check(exitCode === 0, `Check output exit code ${exitCode}`, output);
+      } else {
+        const { exitCode, output } = await run(options.checkCommand, {
+          cwd: project.path,
+          captureExitCode: true,
+          stdio: verbose ? "inherit" : undefined,
+          captureOutput: true,
+        });
+        check(exitCode === 0, `Check output exit code ${exitCode}`, output);
+      }
+    },
+  };
+}
+
+/**
+ * Context passed to the check() function of a package.json check.
+ */
+export interface PackageJsonCheckContext extends CheckContext {
+  /**
+   * JavaScript object representation of the project's package.json file.
+   * This object may be mutated to cause check failures if the mutation results in a change, or to fix the check if in fix mode.
+   */
+  packageJson: PackageJson;
+}
+
+export interface PackageJsonCheckOptions extends CheckOptions {
+  /**
+   * Run the check. The package.json is provided as a JavaScript object at the top level of the context, `packageJson`.
+   * This object may be mutated to cause check failures if the mutation results in a change, or to fix the check if in fix mode.
+   * @param context - context the check is running in
+   */
+  check(context: PackageJsonCheckContext): Promise<void>;
+}
+
+/**
+ * A check which validates properties of the package.json.
+ * In addition to calling check(), the check function may mutate the input packageJson provided in the CheckContext. If it does this,
+ * any material changes to the input packageJson will cause the check to fail, except if in fix mode (and the check has set allowFix to true). In
+ * that case, the check will be fixed by writing the updated package.json to disk.
+ *
+ * @param options options - options for the check
+ * @returns - the created check
+ */
+export function packageJsonCheck(options: PackageJsonCheckOptions): Check {
+  return {
+    ...options,
+    async check({ fix, project, verbose }) {
+      const { packageJson: originalPackageJson } = project;
+
+      // The check may mutate the package.json
+      const packageJsonClone = structuredClone(originalPackageJson);
+      await options.check({
+        verbose,
+        packageJson: packageJsonClone,
+        project: { ...project, packageJson: packageJsonClone },
+        fix,
+      });
+
+      // Check if the cloned package JSON and original are different
+      if (JSON.stringify(originalPackageJson) !== JSON.stringify(packageJsonClone)) {
+        if (fix) {
+          const newPackageJsonContent = await format(JSON.stringify(packageJsonClone), "json");
+          await fs.writeFile(path.join(project.path, "package.json"), newPackageJsonContent);
+        } else {
+          check(false, options.description ?? "package.json changed unexpectedly");
```

Comment:
one downside here is that I can imagine a check function might be able to produce a more meaningful error and link to guidance about what the acceptable values are.